### PR TITLE
Add support for data plane to forward remote IP when given via Proxy Protocol

### PR DESCRIPTION
**Pull request by lfarrel6**
# Why
Services running within a Cage should be able to filter traffic based on the IP of the client

# How
- Update data plane to optionally parse the proxy protocol headers from the socket
- Add proxy protocol supporting tcp and vsock servers to shared lib
- Introduce `ProxiedConnection` trait. 
  - All connections implement the ProxiedConnection trait to be compatible with the generic Listener trait. The default behaviour is to return `None` for the Proxy Protocol header, and `false` for `has_proxy_protocol`
  - The `AcceptedConn` wrapper is created after accepting a connection expected to be proxied. It is the only non-default implementor of the `ProxiedConnection` trait. It returns the values read from the socket (if they were parsed correctly)
- Update data-plane server to write the remote IP and proto into the HTTP headers when available (and TLS Termination is enabled).
- Updated e2e tests in CI to generate certs on the fly to remove hard coded certs which had expired.

### CODE REVIEW

File: shared/src/server/proxy_protocol.rs [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    fn new(
+        inner: S,
+        proxy_protocol: Option<PPHeader<'static>>,
+        opt_preread_buffer: Option<Vec<u8>>,
+    ) -> Self {
+        // if the preread buffer is empty, we don't need to store it
+        let preread_buffer = match opt_preread_buffer.as_deref() {
+            Some(buf) if buf.len() == 0 => None,
+            Some(_) => opt_preread_buffer,
+            None => None,
```
Comment (donaltuohy) [inline shared/src/server/proxy_protocol.rs:23]: nitpick to reduce branches:

```suggestion
            Some(buf) if buf.len() > 0 => buf,
            _ => None,
```

File: shared/src/server/proxy_protocol.rs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+pub async fn try_parse_proxy_protocol<S: AsyncRead + AsyncWrite + Unpin + Sync>(
+    mut incoming_conn: S,
+) -> ServerResult<AcceptedConn<S>> {
+    let mut buf: Vec<u8> = Vec::with_capacity(MIN_PROXY_PROTOCOL_HEADER_LEN);
+    let mut total_read_bytes = 0;
+    loop {
+        let read_len = incoming_conn.read_buf(&mut buf).await?;
+        if read_len == 0 {
+            println!("empty read");
```
Comment (donaltuohy) [inline shared/src/server/proxy_protocol.rs:88]: Looks like there's a few debug logs left in?

**Review by donaltuohy — APPROVED** [review]
Nice lgtm! Just 2 minor things

File: shared/src/server/proxy_protocol.rs (hunk 2 above)
Comment (lfarrel6) [reply shared/src/server/proxy_protocol.rs]: removed!

**Review by donaltuohy — APPROVED** [review]

**Pull request merged**